### PR TITLE
transport sockets: support passthrough state for internal connections

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Implements a special transport socket for transferring state (metadata and filter state) over the internal connection.
This transport socket captures a subset of endpoint metadata, cluster metadata, and stream filter state in the user space socket. When an internal listener accepts a user space socket connection, it immediately merges this passthrough state into the connection stream info. Because the state can be transferred from HTTP stream to TCP stream, this transport socket also participates in the hashing decisions in the HTTP connection pools.

Commit Message: Add passhtrough state over internal connection. 
Risk Level: low, new extension
Testing:  WIP
Docs Changes: yes
Release Notes: yes
Platform Specific Features:
Fixes: #19274

### CODE REVIEW

**Comment by lambdai** [comment]
@kyessenov Should we add the target internal listener address as the first field of this transport socket?

**Comment by kyessenov** [comment]
@lambdai I am not sure since we already have that as `lb_endpoint` address. I think having physical `lb_endpoint` address differ from the internal address in transport socket might be confusing.

**Comment by kyessenov** [comment]
CC @alyssawilk in case you are interested. This is bare minimum needed to propagate policies from the first upstream to the internal listener. but also could be used to propagate the actual destination.

**Comment by alyssawilk** [comment]
@adisuissa ping!

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
I suggest maybe setting the proto as WIP, and add `#not-implemented-hide`, but up to you.

/lgtm api

**Comment by wrowe** [comment]
> I suggest maybe setting the proto as WIP, and add `#not-implemented-hide`

@kyessenov - your thought? In any case, merge master may hopefully clear coverage?

**Comment by kyessenov** [comment]
Thanks, marked as WIP. Implementation likely needs https://github.com/envoyproxy/envoy/pull/19467 first to land.

File: CODEOWNERS
Comment (kyessenov) [inline CODEOWNERS:58]: todo: need a maintainer.

**Comment by wrowe** [comment]
/wait
on WIP...

**Comment by kyessenov** [comment]
@lambdai This is mostly ready implementation-wise, working on the test coverage.

File: source/server/active_internal_listener.cc
Comment (lambdai) [inline source/server/active_internal_listener.cc:57]: We can modify the interface ActiveInternalListener::onAccept(Network::ConnectionSocketPtr&& socket) to
 ActiveInternalListener::onAccept(Network::ConnectionSocketPtr&& socket, UserSpace::IoHandle) to avoid dynamic cast, 
but it doesn't resolve the polluting the deps because IoHandle is in extension.

I am wondering if we can list ActiveInternalListener as an exception for now

**Review by lambdai — COMMENTED** [review]
Flush my comment

File: envoy/network/transport_socket.h
Comment (kyessenov) [reply envoy/network/transport_socket.h:246]: There's a header dependency loop that requires duplicating `using` unfortunately. The symbol is forward-declared.

File: source/extensions/transport_sockets/internal/internal.cc
Comment (kyessenov) [reply source/extensions/transport_sockets/internal/internal.cc:22]: Values after moves are undefined so this is just eliminating potential source of errors.

File: source/server/active_internal_listener.cc
Comment (kyessenov) [reply source/server/active_internal_listener.cc:57]: I don't like this dependency loop. `ActiveInternalListener` is in the core but it only works with an extension `io_socket/user_space`. That shouldn't be allowed. We need to either move both to extension or make `io_socket/user_space` be effectively in core. cc @alyssawilk

File: source/extensions/transport_sockets/internal/config.cc
Comment (kyessenov) [reply source/extensions/transport_sockets/internal/config.cc:118]: Added a metric, will push commit soon.

File: CODEOWNERS
Comment (kyessenov) [reply CODEOWNERS:58]: Yeah, no problem.

File: source/server/active_internal_listener.cc
Comment (alyssawilk) [reply source/server/active_internal_listener.cc:57]: I'd be inclined to make both extensions, especially if Envoy can build without the active internal listener. IMO it ideally would be optional and something folks who don't need can leave out of the binary.

**Comment by wrowe** [comment]
/wait
for unresolved review items, and coverage fix (try merging main)

File: source/server/active_internal_listener.cc
Comment (kyessenov) [reply source/server/active_internal_listener.cc:57]: This is a non-trivial cleanup, filed https://github.com/envoyproxy/envoy/issues/20798.

File: api/envoy/extensions/transport_sockets/internal/v3/internal_upstream.proto
Comment (kyessenov) [reply api/envoy/extensions/transport_sockets/internal/v3/internal_upstream.proto:27]: Ack.

File: api/envoy/extensions/transport_sockets/internal/v3/internal_upstream.proto
Comment (kyessenov) [reply api/envoy/extensions/transport_sockets/internal/v3/internal_upstream.proto:35]: Ack.

File: api/envoy/extensions/transport_sockets/internal/v3/internal_upstream.proto
Comment (kyessenov) [reply api/envoy/extensions/transport_sockets/internal/v3/internal_upstream.proto:58]: Ok, spelled out more details.

File: api/envoy/extensions/transport_sockets/internal/v3/internal_upstream.proto
Comment (kyessenov) [reply api/envoy/extensions/transport_sockets/internal/v3/internal_upstream.proto:43]: Ack.

File: source/common/upstream/upstream_impl.cc
Comment (kyessenov) [reply source/common/upstream/upstream_impl.cc:348]: right, I did move it in the other clause.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Left a few comments (mainly API).

**Comment by wrowe** [comment]
/retest

as simply times out in coverage, incrementally building and testing may be productive.

**Comment by kyessenov** [comment]
Thanks, I'm back from two weeks vacation and will address the comments soon.

File: api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto
Comment (adisuissa) [inline api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto:43]: This field is optional. Please add what's the default value to the comment.

File: api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto
Comment (adisuissa) [inline api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto:52]: Should the field number be 1?

**Review by adisuissa — COMMENTED** [review]
Sorry for the delay, left a few questions.

File: api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto
Comment (kyessenov) [reply api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto:62]: Yes, it will override provided there is source. But it's possible to not have a source metadata. For example:
```
cluster1: Metadata[X=Y]
- endpoint1: Metadata[X=Z]
- endpoint2: -
```

If we specify as source:
- {kind: {cluster: {}}, name: X}
- {kind: {endpoint: {}}, name: X}

Then for endpoint1 the passthrough metadata is Z while for endpoint2 the passthrough metadata is Y.

I think your suggestion makes sense at runtime, we can resolve overrides at extraction.

File: api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto
Comment (kyessenov) [reply api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto:69]: Yes, it's reasonable not to have it. For example, a TCP-only internal listener that only relies on endpoint metadata to define the tunnel endpoint. This state is useful when there are custom filters on both downstream and internal listeners, that collect telemetry, for instance.

**Comment by wrowe** [comment]
Ping; @kyessenov please tag resolved discussions you trust are solved, and @adisuissa please check that this meets api requirements. Thanks all!

File: source/server/active_internal_listener.cc
Comment (kyessenov) [reply source/server/active_internal_listener.cc:57]: I'll address the dependency loop separately. It requires some core changes that shouldn't be coupled with this PR.

File: api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto
Comment (adisuissa) [inline api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto:60]:
```suggestion
  // locations, and both locations contain the metadata with the given name,
```

**Review by adisuissa — COMMENTED** [review]
Thanks!
API lgtm modulo a small nit.
Looked at old comments on this PR, and saw that at one point you thought of making the API WIP. If this is still the case, the annotation is missing.
If this feature is fully implemented, I suggest adding/updating the docs and release notes with the relevant info.

File: api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto
Comment (kyessenov) [reply api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto]: Done.

**Review by adisuissa — COMMENTED** [review]
Thanks!
/lgtm api

**Comment by wrowe** [comment]
@alyssawilk or @antoniovicente could you provide the final review, as Sr. Maintainers? This looks like it's across the finish line.

**Comment by alyssawilk** [comment]
If Lizan's going to co-own I think he should do senior maintainer reviews, but I do have a drive by question/comment while I'm looking :-)

**Comment by kyessenov** [comment]
I haven't asked @lizan directly, so please feel free to re-assign another co-owner if necessary.

**Comment by kyessenov** [comment]
@alyssawilk Did you forget to post the drive-by question :) ?

File: source/common/upstream/upstream_impl.cc
Comment (kyessenov) [reply source/common/upstream/upstream_impl.cc:341]: Hmm, even if we add `createTransportSocket` to take a host, if the options are shared per cluster, that won't work, right?    I noticed that there is a metadata matcher for transport sockets, which is implemented at cluster level, which is also aware of per-host metadata, but that's about it, every other transport socket does not need host access.

The alternative might be to make the upstream host available in the callbacks or context for transport sockets. I can explore that, but I suspect to have some place to hold the shared host pointer, it'll be a lot of glue, too.

Comment (kyessenov) [reply source/common/upstream/upstream_impl.cc:341]: Actually, we already capture filter state in transport socket options. So the options have to be per-request (owned by HCM instance?) and cannot be shared unless I misunderstand something.

File: source/extensions/transport_sockets/internal_upstream/config.cc
Comment (soulxu) [inline source/extensions/transport_sockets/internal_upstream/config.cc:118]: Is it possible to not copy those filter state data and just pass the filter_state reference to the downstream listener? since we invoke the downstream listener's onAccept immediately when create the client connection (even we didn't return the client connection yet), 
https://github.com/envoyproxy/envoy/blob/006bbc3614724ead239fcc3a2438b4dd6b9173e6/source/extensions/bootstrap/internal_listener/client_connection_factory.cc#L66

I'm thinking is that means the upstream stream info and state_filter are still alive and have no chance to be released.

File: source/extensions/transport_sockets/internal_upstream/internal_upstream.cc
Comment (kyessenov) [reply source/extensions/transport_sockets/internal_upstream/internal_upstream.cc:18]: Yes, we want this to be contained in the internal listener extensions. It's not clear how to pass C++ state across true POSIX IO handles in general.

File: source/extensions/transport_sockets/internal_upstream/config.h
Comment (kyessenov) [reply source/extensions/transport_sockets/internal_upstream/config.h:19]: Yes, thanks, will do.

File: source/extensions/transport_sockets/internal_upstream/internal_upstream.cc
Comment (soulxu) [reply source/extensions/transport_sockets/internal_upstream/internal_upstream.cc:18]: It seems ok for me just add the `passthroughState` interface to Network::IoHandle, but POSIX and Windows IO handles just not implement it and return `nullptr` directy.

**Comment by wrowe** [comment]
There are several open conversations, so we'll patiently...

/wait

File: source/extensions/transport_sockets/internal_upstream/internal_upstream.cc
Comment (kyessenov) [reply source/extensions/transport_sockets/internal_upstream/internal_upstream.cc:18]: I don't think I want to impose this passthrough state on all sockets yet, but it's certainly something we can add later.

File: source/common/network/transport_socket_options_impl.h
Comment (kyessenov) [reply source/common/network/transport_socket_options_impl.h]: Stale.

File: source/common/upstream/upstream_impl.cc
Comment (kyessenov) [reply source/common/upstream/upstream_impl.cc]: Accepted the suggestion, but it cost me in touching many call sites.

File: source/extensions/transport_sockets/internal_upstream/config.cc
Comment (kyessenov) [reply source/extensions/transport_sockets/internal_upstream/config.cc:118]: What might be confusing - this is downstream stream info, not upstream stream info. I didn't think about transfer from downstream to upstream, which also makes sense, but is not needed here.

File: source/extensions/transport_sockets/internal_upstream/internal_upstream.cc
Comment (kyessenov) [reply source/extensions/transport_sockets/internal_upstream/internal_upstream.cc:18]: I've looked and IoHandle in user space and IoHandle in the envoy network namespace diverged quite a lot. E.g. there is no`recv` or  `sendmsg` in user space one. So I'm not sure how to promote the user space IoHandle to base interface. Thoughts? @lambdai

**Comment by kyessenov** [comment]
I think I answered most of the comments. Sorry this is taking too long. Can we assign some reviewer for the implementation? It doesn't have to be @lizan .

File: envoy/network/transport_socket.h
Comment (lizan) [inline envoy/network/transport_socket.h:265]: It is undesired that transport socket will have dependency on upstream while this interface is not specific to upstream. Since we only need filter state and metadata here, why not use the filterState in TransportSocketOptions?

Comment (kyessenov) [reply envoy/network/transport_socket.h:265]: The filter state is already in the options, so supporting just the filter state does not need any interface changes.

How would I propagate the endpoint metadata to the transport socket? I don't think we can write a downstream filter easily that copies metadata into the filter state, since the upstream owns the host selection.

File: source/extensions/transport_sockets/internal_upstream/config.cc
Comment (soulxu) [reply source/extensions/transport_sockets/internal_upstream/config.cc:118]: What I'm thinking is utilizing the special part of the internal connection. When the downstream creates the internal `Network::ClientConnectionImpl` by `InternalClientConnectionFactory::createClientConnection`, it will invoke the upstream's `onAccept` callback directly inside `InternalClientConnectionFactory::createClientConnection`, 
https://github.com/envoyproxy/envoy/blob/006bbc3614724ead239fcc3a2438b4dd6b9173e6/source/extensions/bootstrap/internal_listener/client_connection_factory.cc#L66

so I'm thinking that means the downstream and upstream lifetime already bind together. When you do the merge with upstream filter states in the InternalListener's `onAccept` callback, the downstream filter stats shouldn't have chance to end the life, because the `InternalClientConnectionFactory::createClientConnection` method call event not finish yet. (onAccept call is the last line of this method)

https://github.com/envoyproxy/envoy/pull/19435/files#diff-e0f716749375fc91f21d624aca995ad6a0af4147666fee56b9703e8774e2dc83R64

But I didn't validate it, I could be wrong.

File: envoy/network/transport_socket.h
Comment (lizan) [reply envoy/network/transport_socket.h:265]: Yeah that actually makes more sense.

Comment (kyessenov) [reply envoy/network/transport_socket.h:265]: SG. I'd like to split the refactor to minimize code changes. It seems to touch tons of files. Will push a separate PR to split downstream and upstream transport sockets.

Comment (kyessenov) [reply envoy/network/transport_socket.h:265]: Filed https://github.com/envoyproxy/envoy/pull/21648, slogging through test build fixes.

**Comment by lizan** [comment]
@kyessenov I merged https://github.com/envoyproxy/envoy/pull/21648, can you update this PR?

/wait

**Comment by kyessenov** [comment]
/retest
(timeout)

**Comment by wrowe** [comment]
Build looks clean, @lizan - are your concerns addressed?

**Review by lizan — APPROVED** [review]

**Comment by ggreenway** [comment]
@kyessenov there's a merge conflict to resolve.

/wait

**Comment by kyessenov** [comment]
Thanks, fixed the conflict.

**Review by lizan — APPROVED** [review]

**Comment by kyessenov** [comment]
/retest
(possible flake //test/extensions/access_loggers/open_telemetry:grpc_access_log_impl_test)

**Comment by phlax** [comment]
> It's a new flake caused by https://github.com/envoyproxy/envoy/commit/f7a8db0599924663a3549460edeb212be97e2ced.

apologies, addressing this now

**Pull request merged**
